homeassistant/components/tellduslive/__init__.py

```diff
@@ -0,0 +1,312 @@
+
+    dev_reg = await hass.helpers.device_registry.async_get_registry()
+    for hub in await client.get_hubs():
+        _LOGGER.debug("Connected hub %s", hub['name'])
+        dev_reg.async_get_or_create(
+            config_entry_id=entry.entry_id,
+            connections={('IP', hub['ip'])},
+            identifiers={(DOMAIN, hub['id'])},
+            manufacturer='Telldus',
+            name=hub['name'],
+            model=hub['type'],
+            sw_version=hub['version'],
+        )
+
+    await client.update()
+
+    return True
+
+
+async def async_setup(hass, config):
+    """Set up the Telldus Live component."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={'source': config_entries.SOURCE_IMPORT},
+            data={KEY_HOST: config.get(DOMAIN, {}).get(CONF_HOST)}))
+
+    return True
+
+
+class TelldusLiveClient:
+    """Get the latest data and update the states."""
+
+    def __init__(self,
+                 hass,
+                 session,
+                 config_entry,
+                 interval=DEFAULT_UPDATE_INTERVAL):
+        """Initialize the Tellus data object."""
+        self._known_devices = set()
+        self.entities_info = {}
+        self._hass = hass
+        self._config_entry = config_entry
+        self._client = session
+
+        async_track_time_interval(self._hass, self.update,
+                                  DEFAULT_UPDATE_INTERVAL)
+
+    async def get_hubs(self):
+        """Return hubs registered for the user."""
+        return self._client.get_clients()
+
+    async def update(self, *args):
+        """Periodically poll the servers for current state."""
+        _LOGGER.debug('Updating')
+        await self._sync()
+
+    async def _sync(self):
+        """Update local list of devices."""
+        if not self._client.update():
+            _LOGGER.warning('Failed request')
+
+        def identify_device(device):
+            """Find out what type of HA component to create."""
+            from tellduslive import (DIM, UP, TURNON)
+            if device.methods & DIM:
+                return 'light'
+            if device.methods & UP:
+                return 'cover'
+            if device.methods & TURNON:
+                return 'switch'
+            if device.methods == 0:
+                return 'binary_sensor'
+            _LOGGER.warning("Unidentified device type (methods: %d)",
+                            device.methods)
+            return 'switch'
+
+        async def discover(device_id, component):
+            """Discover the component."""
+            config_entries_key = '{}.{}'.format(component, DOMAIN)
+            async with self._hass.data[DATA_CONFIG_ENTRY_LOCK]:
+                if config_entries_key not in self._hass.data[
+                        CONFIG_ENTRY_IS_SETUP]:
+                    await self._hass.config_entries.async_forward_entry_setup(
+                        self._config_entry, component)
```

Comment:
@MartinHjelmare is it a good idea to dynamically load entry platforms like this or would you prefer to always load all platforms (I guess that we might save some memory by not having a bunch of platforms loaded that just are waiting for new entities)?

(This is stolen from the MQTT component)

Reply:
That the components are loaded in the `init`stage (or even the setup stage).

Reply:
But we would need to forward entry setup whenever a new device is detected, right? So we still need this. I'm not really seeing what the alternative would be.

Reply:
No, not really. There is no way to pass the id of the new entity so I'm using a `dispatch` to send the new id.

```python
async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up tellduslive sensors dynamically."""
    async def async_discover_switch(device_id):
        """Discover and add a discovered sensor."""
        client = hass.data[DOMAIN][config_entry.entry_id]
        async_add_entities([TelldusLiveSwitch(client, device_id)])
     async_dispatcher_connect(
        hass, TELLDUS_DISCOVERY_NEW.format(switch.DOMAIN, DOMAIN),
        async_discover_switch)
```

Reply:
If I were to return `async_add_entities` from `async_setup_entry` I could call that from my platform when new devices are found.

Reply:
Here is the code from `mqtt`

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mqtt/discovery.py#L237

Reply:
Perhaps, but that will have to be someone else's job